### ISSUE TITLE
fake address space separation for DPMI

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#622]
As discussed here:
https://github.com/stsp/dosemu2/issues/612#issuecomment-393529530
Should probably wait for #620, after which
it would be really very simple to implement.

**Comment by stsp** [comment]
So this is only needed to allow
multiple clients to do fixed linear
allocs. Not sure if worth a trouble.
As for mapping, this means 32
separate mapping windows (1 per
possible DPMI client). Will likely
also complicate the remote DPMI
hook.

**Comment by stsp** [comment]
Actually @Baron-von-Riedesel can
you fix this yourself? I think you
can avoid using fixed allocs, and
do normal allocs instead. Then
you can just set the bases on all
LDT descriptors accordingly, and
get the desired 4Mb offset.

In fact, I think the best strategy
is to preallocate the large pool
of uncommitted pages, set the
segments, and do the memory
management on your own by
changing page attributes and
using a custom allocator. This
will make you HX compatible
with any 0.9 DPMI.

Address space separation on
dosemu2 side is likely not a
deal, and not the right thing
to do.

What do you think?

**Comment by Baron-von-Riedesel** [comment]
> Actually @Baron-von-Riedesel can you fix this yourself? 

Probably yes. I guess it's not possible to map DosEmu's emulated VGA device memory at the "faked" 0xA0000-0xBffff range - but that's a minor problem, as far as text mode apps are concerned.

**Comment by stsp** [comment]
> I guess it's not possible to map DosEmu's emulated VGA device memory at the "faked"

Well, unless you can use the
dedicated segreg for that
(like %gs), you likely can use
fns 508 (map device) and 509
(map conv mem).
Unfortunately neither is
implemented correctly under
dosemu2, and in fact they are
optional. But if you want to
use either one, I can implement
it once you provide the test-case.

**Comment by stsp** [comment]
> Unfortunately neither is
> implemented correctly under
> dosemu2

I guess I was completely
misunderstanding fn 509,
which is documented to
only map the previously
allocated (by current client)
DOS memory blocks...
In this case it is actually
implemented properly
under dosemu2, but is
not suitable for the
aforementioned task.

How about implementing
fn 508 and map entire 1st
MB of physical ram to every
client? This would be exactly
what I though fn 509 is, and
since it appears not, I feel
like we really need 508 to
be implemented.
It will not be the exact copy
of 1st MB, as it will bypass
EMS and will ignore a20 wrap,
but the physical layout will
be identical, which means
VGA.

**Comment by Baron-von-Riedesel** [comment]
>... once you provide the test-case.

Here's a very simple test case. allocs a 4MB+x block, uncommited, and tries to map 
1. DOS memory at offset 0, size 640kB
2. VGA memory at offset 0xb8000, size 32 kB
3. "normal" memory at offset 0x400000, size x

Then it uses DPMI func 501h to alloc a memory block.
Output should look like:
```
int 31h, ax=6 for CS returned CX:DX=0
int 31h, ax=6 for SS returned CX:DX=0
int 31h, ax=501h returned BX:CX=408000, SI:DI=518000
reading addr 408000...ok
```
(the value behind "SI:DI" is specific for HDPMI)

[mem.exe.zip](https://github.com/user-attachments/files/19723303/mem.exe.zip)

**Comment by stsp** [comment]
Could you please add sources?

**Comment by Baron-von-Riedesel** [comment]
> Could you please add sources?

Yes, but the stub LOADPEVM.BIN is part of HX - you won't be able to create the binary without a few include files from the HX source tree. And the stub is the interesting part of it, since it setups the "VM".

[STUBVM.zip](https://github.com/user-attachments/files/19724690/STUBVM.zip)

**Comment by stsp** [comment]
So you decided to use both
508 and 509, but please note
that 509 is documented this
way:
```
A client may only map conventional memory that it already owns; i.e. memory which the client previously
allocated with Int 31H Function 0100H or by calling
DOS's Int 21H Function 48H directly via the translation services.
```
So you have the similar
misunderstanding of 509
as I had before reading the
actual spec.
Another approach may be
to map 1st mb entirely with
508, but I won't push into
that direction since it also
looks problematic.
I wish 509 could be used for
mapping entire lowmem,
including VGA, EMS and HMA
(and forget about 508),
but its just not the case. :(

So for now I'll just try to
comply with your test-case
and not with the specs.
It would still be a great relief
to drop those 4mb reservations
required by current HX code,
so whatever spec says, can be
sacrificed to that.

**Comment by Baron-von-Riedesel** [comment]
> So you have the similar misunderstanding of 509 as I had before reading the actual spec.

I guess I once read the spec and choose to ignore the limitations - after all, there's no "gold standard", meaning a real implementation of v1.0 by a known vendor. And the specs don't even mention the problem that - as far as DOS memory is concerned - there have to be significant differences between DPMI hosts with an integrated v86-monitor ( like Windows and DosEmu ) and DPMI hosts that run as VCPI clients (hdpmi ).

**Comment by stsp** [comment]
> I guess I once read the spec and choose to ignore the limitations

In this case the one could
assume to map entire 1st
MB with 509, including
EMS, VGA and HMA.
You took the inconsistent
approach that can map
some things but not others.
But I agree this probably is
the most practical extension
for this particular task.
Lets hope DPMI clients are
not going to access the EMS
frame via mapped space.

> there have to be significant differences between DPMI hosts

Which ones?
I think you can do mostly
the same with RM as with v86.

**Comment by Baron-von-Riedesel** [comment]
> Which ones?

Well, mostly the one you mentioned yourself: "Lets hope DPMI clients are not going to access the EMS frame via mapped space."
It's not just the EMS page frame, as with EMS v4.0 huge parts of conventional memory can be remapped at any time. There's no way to inform DPMI hosts running as VCPI clients of such changes.

**Comment by stsp** [comment]
Firstly I am a bit confused by
your differentiation: DPMI
hosts running as VCPI clients
do _also_ use v86 monitor.
Maybe you mean the DPMI
hosts started out of real-mode?

Secondly I do not see how
would you implement EMS
for 509 in either case - by
trapping EMS calls? I am not
aware of a sufficiently simple
way of mapping EMS into
DPMI space, be it with v86
or VCPI or whatever.

**Comment by Baron-von-Riedesel** [comment]
> DPMI hosts running as VCPI clients do also use v86 monitor.

Yes, but they don't know the status of the first MB's page table when they switch to v86-mode (via VCPI).

> Secondly I do not see how would you implement EMS for 509 in either case 

The DPMI host doesn't care about EMS - but it may need to copy some data to conventional memory and for this copy operation it uses the paging table it once got from VCPI ( int 67h, ax=DE01h ) - and this table may have been changed in the meantime by some EMS v4.0 calls.

**Comment by stsp** [comment]
Should now be fixed, enjoy!
Please make a note here or
in discussions if/when you
can deploy the new HX code,
as supporting the current
versions is a PITA.

> The DPMI host doesn't care about EMS

As long as 509 is not allowing
to map the EMS frame. What
I meant by "Lets hope DPMI
clients are not going to access
the EMS frame via mapped space."
was a bit more complex. :)
I meant that to make fake
address spaces consistent,
you may need to map 1st MB
to every fake address space.
Then you very much recreate
the usual "unaliased" layout.
However some DPMI client
may try to access EMS, and
since it doesn't know about
the fake address spaces, he
may just set up the LDT for
e0000. But since you fake,
you'll add the offset to such
LDT. And now the client can
mostly see the 1st MB, but
EMS doesn't work for him!
